Attribute support

m-vo opened the pull request:
This PR adds attribute support which basically allows doing this on PHP>=8:

```diff
- /**
-  * @ServiceTag("target_service", priority=123, bar="baz")
-  */
+ #[ServiceTag('target_service', priority: 123, bar: 'baz')]
class Foo
{
}
```

(This also adds phpunit for testing - not sure about your preferences herewith for this bundle. :slightly_smiling_face: )

m-vo committed:
add + configure phpunit

m-vo committed:
add \Attribute to ServiceTag class and introduce BC layer in constructor

m-vo committed:
parse attributes in ServiceAnnotationPass

m-vo committed:
add tests

m-vo committed:
add attributes example to README

m-vo committed:
include repeated annotations/attributes in test-cases

aschempp commented:
very good! 👏 

I'm not very experienced with attributes, wouldn't it make sense to use a separate class for attributes?

m-vo commented:
Both would work I guess. There was also a similar discussion in https://github.com/symfony/symfony/pull/37474 for the `Route` attribute. There, now both strategies share a single class: https://github.com/symfony/symfony/blob/896f4a6eb6a986cc29e096dfe1f3f7042b03126f/src/Symfony/Component/Routing/Annotation/Route.php#L14-L26

I like that you can easily refactor your code (not even a new import needed) and the behavior stays identical. On the downside there is less type safety in the constructor. But IMHO this could/should be tweaked in `AnnotationReader#getClassAnnotations()` if this becomes a problem. :thinking: .

leofeyer commented on src/Annotation/ServiceTag.php:
```diff
@@ -22,6 +22,7 @@
  *     @Attribute("value", required = true, type = "string"),
  * })
  */
+#[\Attribute(\Attribute::IS_REPEATABLE | \Attribute::TARGET_CLASS | \Attribute::TARGET_METHOD)]
```
Why the leading backslash? Is there an `Attribute` class in the global namespace as well?

Toflar replied:
What do you mean with "as well". Yes, that's the core PHP8 implementation of attributes :)

leofeyer replied:
By "as well" I mean that there is another `Attribute` class imported above (`use Doctrine\Common\Annotations\Annotation\Attribute`). But I realize that this one is used for the annotations.

Toflar commented on phpunit.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
That should be `phpunit.xml.dist` :)

m-vo replied:
doen in ae3be49c881ff180b718e6ccadb6599919426c57

Toflar reviewed (changes requested):
Looks great! Only a minor nitpick about file naming :)

m-vo committed:
rename phpunit.xml to phpunit.xml.dist

ausi commented:
Did you consider using named arguments instead of the array?

```php
#[ServiceTag('target_service', priority: 123, bar: 'baz')]
```

Should work fine using the constructor signature:

```php
__construct($name, ...$data)
```

m-vo commented:
> Did you consider using named arguments instead of the array?

Didn't know that this works with arbitrary names. :relaxed: 
I actually like this! Maybe it's even possible to specify type hints for these things in the future.

@Toflar @aschempp Wdyt?

m-vo committed:
use named arguments instead of an array of data

m-vo commented:
Switched to named arguments in 19c6bf04ca4cf706e02725d157c4522a0dd9af19.

I also renamed the first argument to `$serviceName` as it's less likely to clash with any of the named data arguments (in contrast to `$value` or `$name`).

aschempp commented:
Very nice! I also didn't know about that 🙃 

> I also renamed the first argument to $serviceName as it's less likely to clash with any of the named data arguments (in contrast to $value or $name).

I would actually go with `$value` here, since that's the internal name the annotation reader uses as well, and having a `value` key on a tag wouldn't work with that either. WDYT?

m-vo commented:
IMHO the downside to `$value` is that it doesn't make it obvious what to enter when you're using auto completion in your IDE.

But I'm fine with both.

aschempp commented:
That's true. But `$serviceName` doesn't make that either? Wouldn't it be `$tagName` or just `$name` then (similar to YAML `{ name: kernel.event_listener, event: foobar }`? Also consider that we have child annotations/attributes in Contao where the value has a different meaning (but they can use a different constructor, so that might not be relevant).

aschempp commented:
@Toflar just pointed me to https://symfony.com/blog/new-in-symfony-5-3-service-autoconfiguration-and-attributes
maybe that whole bundle will become obsolete?

m-vo commented:
> maybe that whole bundle will become obsolete?

Not sure about this. If I get this correctly, you could get rid of the `ServiceTagInterface` but would still need the plumbing like so:
```php
$container->registerAttributeForAutoconfiguration(
  ServiceTag::class, 
  static function (ChildDefinition $definition, ServiceTag $attribute): void {
    $definition->addTag($attribute->getName(), $attribute->getAttributes());
  }
);
```

I guess Contao (or any consumer) could implement their special attributes without this bundle but would then on the other hand need to require the Sf DI component in >= 5.3. Hmm.

(Speaking of Contao: We should consider following the `As<Something>` naming convention of Symfony. See https://github.com/symfony/symfony/pull/40556…)

aschempp commented:
yes that's what I mean. Instead of adding attributes trough this bundle, we could add support for attributes when Contao requires Symfony 5.3 (like in 4.12 or 4.13). If we agree we MUST have attribute support e.g. in 4.9, then we need to implement it in this bundle.

aschempp commented:
Attributes are supported natively by Symfony and Contao, therefore I'm closing this PR.
